submission/persistence-mongo/src/main/kotlin/ac/uk/ebi/biostd/persistence/doc/service/StatsMongoDataService.kt

```diff
@@ -0,0 +1,66 @@
+package ac.uk.ebi.biostd.persistence.doc.service
+
+import ac.uk.ebi.biostd.persistence.common.exception.StatNotFoundException
+import ac.uk.ebi.biostd.persistence.common.model.SubmissionStat
+import ac.uk.ebi.biostd.persistence.common.model.SubmissionStatType
+import ac.uk.ebi.biostd.persistence.common.request.PaginationFilter
+import ac.uk.ebi.biostd.persistence.common.service.StatsDataService
+import ac.uk.ebi.biostd.persistence.doc.db.data.SubmissionDocDataRepository
+import ac.uk.ebi.biostd.persistence.doc.model.DocStat
+import ac.uk.ebi.biostd.persistence.doc.model.DocSubmission
+import ac.uk.ebi.biostd.persistence.doc.model.SingleSubmissionStat
+import org.springframework.data.domain.PageRequest
+
+class StatsMongoDataService(
+    private val subDataRepository: SubmissionDocDataRepository
+) : StatsDataService {
+    override fun findByType(submissionStatType: SubmissionStatType, filter: PaginationFilter): List<SubmissionStat> =
+        subDataRepository
+            .findAllByStatType(submissionStatType, PageRequest.of(filter.pageNumber, filter.limit))
+            .content
+            .map { toSubmissionStat(it, submissionStatType) }
+
+    override fun findByAccNoAndType(accNo: String, submissionStatType: SubmissionStatType): SubmissionStat =
+        subDataRepository
+            .findByAccNoAndStatType(accNo, submissionStatType)
+            ?.let { toSubmissionStat(it, submissionStatType) }
+            ?: throw StatNotFoundException(accNo, submissionStatType)
+
+    override fun saveAll(stats: List<SubmissionStat>): List<SubmissionStat> =
+        stats.filterValid()
+            .mapValues { it.value.last() }
+            .map { updateSubmissionStats(it.key!!, it.value) }
+
+    override fun incrementAll(stats: List<SubmissionStat>): List<SubmissionStat> =
+        stats.filterValid()
+            .mapValues { summarizeIncrements(it.key!!.stats, it.value) }
+            .map { updateSubmissionStats(it.key!!, it.value) }
+
+    private fun List<SubmissionStat>.filterValid() =
+        groupBy { it.accNo }
+        .mapKeys { subDataRepository.findByAccNo(it.key.toUpperCase()) }
+        .filterKeys { it != null }
+
+    private fun updateSubmissionStats(sub: DocSubmission, update: SubmissionStat): SubmissionStat {
+        subDataRepository.updateStats(sub.accNo, sub.version, updatedStats(update, sub.stats))
+        return update
+    }
+
+    private fun updatedStats(update: SubmissionStat, currentStats: List<DocStat>): List<DocStat> =
+        currentStats
+            .filterNot { it.name == update.type.name }
+            .plus(DocStat(update.type.name, update.value))
+
+    private fun summarizeIncrements(currentStats: List<DocStat>, increments: List<SubmissionStat>): SubmissionStat {
```

Comment:
not following what is the purpose of this function. Name/code is not self-explanatory I suspect is because is not single purposed. May you explain the logic, please.

Reply:
A file may contain several increments for the same submission, for example:

```
S-BSST1    8
S-BSST2    4
S-BSST1    2
```

The method should summarize all the increments for the same submission so at the end it's incremented the corresponding amount, 10 for S-BSST1, 4 for S-BSST2. The purpose of this method is to summarize this before and then make a single update.

Reply:
I see I think we should perform this in the database so it thread-safe, simpler and efficient.

```
BulkOperations ops = template.bulkOps(BulkMode.UNORDERED, Match.class);
for (val stat : newStats) {
    ops.updateOne(query(where("acc").is(stat.acc).and("type").is(stats.type), createUpdateSet(stat));
}
ops.execute();
```

Reply:
Done